adapter/pkg/metrics/metrics.go

```diff
@@ -0,0 +1,210 @@
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/procfs"
+	"github.com/shirou/gopsutil/v3/cpu"
+	"github.com/shirou/gopsutil/v3/host"
+	"github.com/shirou/gopsutil/v3/load"
+	"github.com/shirou/gopsutil/v3/mem"
+	logger "github.com/wso2/product-microgateway/adapter/internal/loggers"
+	"github.com/wso2/product-microgateway/adapter/pkg/logging"
+)
+
+var (
+	prometheusMetricRegistry = prometheus.NewRegistry()
+
+	hostInfo = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "host_info",
+		Help: "Host Info",
+	}, []string{"os"})
+
+	availableCPUs = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "os_available_cpu_total",
+		Help: "Number of available CPUs.",
+	})
+
+	freePhysicalMemory = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "os_free_physical_memory_bytes",
+		Help: "Amount of free physical memory.",
+	})
+
+	totalVirtualMemory = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "os_total_virtual_memory_bytes",
+		Help: "Amount of total virtual memory.",
+	})
+	usedVirtualMemory = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "os_used_virtual_memory_bytes",
+		Help: "Amount of used virtual memory.",
+	})
+
+	systemCPULoad = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "os_system_cpu_load_percentage",
+		Help: "System-wide CPU usage as a percentage.",
+	})
+
+	loadAvg = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "os_system_load_average",
+		Help: "Current load of CPU in the host system for the last {x} minutes",
+	}, []string{"duration"})
+
+	processStartTime = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "process_start_time_seconds",
+		Help: "Start time of the process since unix epoch in seconds.",
+	})
+
+	processOpenFDs = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "process_open_fds",
+		Help: "Number of open file descriptors.",
+	})
+)
+
+func init() {
+	// Register the Go collector with the registry
+	goCollector := collectors.NewGoCollector()
+	prometheusMetricRegistry.MustRegister(goCollector)
+
+	// Register other metrics
+	prometheusMetricRegistry.MustRegister(hostInfo, availableCPUs, freePhysicalMemory, usedVirtualMemory, totalVirtualMemory,
+		systemCPULoad, loadAvg, processStartTime, processOpenFDs)
+}
+
+// recordMetrics record custom golang metrics
+var recordMetrics = func(collectionInterval int32) {
+	for {
+		host, err := host.Info()
+		if handleError(err, "Failed to get host info") {
+			return
+		}
+		hostInfo.WithLabelValues(host.OS).Set(1)
+		availableCPUs.Set(float64(runtime.NumCPU()))
+
+		v, err := mem.VirtualMemory()
+		if handleError(err, "Failed to read virtual memory metrics") {
+			return
+		}
+		freePhysicalMemory.Set(float64(v.Free))
+		usedVirtualMemory.Set(float64(v.Used))
+		totalVirtualMemory.Set(float64(v.Total))
+
+		percentages, err := cpu.Percent(0, false)
+		if handleError(err, "Failed to read cpu usage metrics") || len(percentages) == 0 {
+			return
+		}
+		systemCPULoad.Set(percentages[0])
```

Comment:
Do we only want the percentage of CPU 1?

Reply:
Fixed in [7b1668e](https://github.com/wso2/product-microgateway/pull/3428/commits/7b1668e3349545785f7702713c6e908a0e79b89f)